{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, \"/u/home/m/michaelw/project-ngarud/microbiome_evolution/py3.8/microbiome_evolution_SHALON/postprocessing_scripts/\")\n",
    "\n",
    "import config\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import subprocess\n",
    "\n",
    "import pickle\n",
    "\n",
    "import os\n",
    "\n",
    "import itertools\n",
    "\n",
    "import datetime\n",
    "\n",
    "#MIDAS postprocessing scripts\n",
    "from calculate_intersample_changes import *\n",
    "import parse_midas_data\n",
    "import diversity_utils\n",
    "import core_gene_utils\n",
    "import parse_patric\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "sns.set_style(\"whitegrid\")\n",
    "\n",
    "within_host_changes = True\n",
    "\n",
    "%matplotlib inline\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_list_path = \"/u/project/ngarud/Garud_lab/metagenomic_fastq_files/Shalon_2023/metadata/species_snps.txt\"\n",
    "\n",
    "with open(species_list_path) as f:\n",
    "    species_list = [line.strip() for line in f if line.strip()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Acidaminococcus_intestini_54097',\n",
       " 'Actinomyces_graevenitzii_58300',\n",
       " 'Actinomyces_sp_57735',\n",
       " 'Actinomyces_sp_62581',\n",
       " 'Actinomyces_viscosus_57672',\n",
       " 'Adlercreutzia_equolifaciens_60310',\n",
       " 'Aggregatibacter_aphrophilus_58143',\n",
       " 'Akkermansia_muciniphila_55290',\n",
       " 'Alistipes_finegoldii_56071',\n",
       " 'Alistipes_indistinctus_62207',\n",
       " 'Alistipes_onderdonkii_55464',\n",
       " 'Alistipes_putredinis_61533',\n",
       " 'Alistipes_senegalensis_58364',\n",
       " 'Alistipes_shahii_62199',\n",
       " 'Anaerostipes_hadrus_55206',\n",
       " 'Atopobium_parvulum_59960',\n",
       " 'Atopobium_sp_59401',\n",
       " 'Bacteroidales_bacterium_58650',\n",
       " 'Bacteroides_caccae_53434',\n",
       " 'Bacteroides_cellulosilyticus_58046',\n",
       " 'Bacteroides_clarus_62282',\n",
       " 'Bacteroides_coprocola_61586',\n",
       " 'Bacteroides_eggerthii_54457',\n",
       " 'Bacteroides_finegoldii_57739',\n",
       " 'Bacteroides_fragilis_54507',\n",
       " 'Bacteroides_intestinalis_61596',\n",
       " 'Bacteroides_massiliensis_44749',\n",
       " 'Bacteroides_ovatus_58035',\n",
       " 'Bacteroides_pectinophilus_61619',\n",
       " 'Bacteroides_plebeius_61623',\n",
       " 'Bacteroides_salyersiae_54873',\n",
       " 'Bacteroides_stercoris_56735',\n",
       " 'Bacteroides_thetaiotaomicron_56941',\n",
       " 'Bacteroides_uniformis_57318',\n",
       " 'Bacteroides_vulgatus_57955',\n",
       " 'Bacteroides_xylanisolvens_57185',\n",
       " 'Barnesiella_intestinihominis_62208',\n",
       " 'Bifidobacterium_adolescentis_56815',\n",
       " 'Bifidobacterium_angulatum_45832',\n",
       " 'Bifidobacterium_animalis_58116']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_list[:40]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "good_species = []\n",
    "\n",
    "for species in species_list:\n",
    "    haploid_samples = diversity_utils.calculate_haploid_samples(species)\n",
    "    if len(haploid_samples) > 1:\n",
    "        good_species.append(species)\n",
    "    else:\n",
    "        continue"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Sample dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_map = parse_midas_data.parse_sample_metadata_map()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing Acidaminococcus_intestini_54097 (1/200)\n",
      "Processing Actinomyces_graevenitzii_58300 (2/200)\n",
      "Processing Actinomyces_sp_57735 (3/200)\n",
      "Processing Actinomyces_sp_62581 (4/200)\n",
      "Processing Actinomyces_viscosus_57672 (5/200)\n",
      "Processing Adlercreutzia_equolifaciens_60310 (6/200)\n",
      "Processing Aggregatibacter_aphrophilus_58143 (7/200)\n",
      "Processing Akkermansia_muciniphila_55290 (8/200)\n",
      "Processing Alistipes_finegoldii_56071 (9/200)\n",
      "Processing Alistipes_indistinctus_62207 (10/200)\n",
      "Processing Alistipes_onderdonkii_55464 (11/200)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_33089/2722743874.py:17: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison\n",
      "  haploid_boolean_vec = [True if sample in haploid_samples else False for sample in high_coverage_samples]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing Alistipes_putredinis_61533 (12/200)\n",
      "Processing Alistipes_senegalensis_58364 (13/200)\n",
      "Processing Alistipes_shahii_62199 (14/200)\n",
      "Processing Anaerostipes_hadrus_55206 (15/200)\n",
      "Processing Atopobium_parvulum_59960 (16/200)\n",
      "Processing Atopobium_sp_59401 (17/200)\n",
      "Processing Bacteroidales_bacterium_58650 (18/200)\n",
      "Processing Bacteroides_caccae_53434 (19/200)\n",
      "Processing Bacteroides_cellulosilyticus_58046 (20/200)\n",
      "Processing Bacteroides_clarus_62282 (21/200)\n",
      "Processing Bacteroides_coprocola_61586 (22/200)\n",
      "Processing Bacteroides_eggerthii_54457 (23/200)\n",
      "Processing Bacteroides_finegoldii_57739 (24/200)\n",
      "Processing Bacteroides_fragilis_54507 (25/200)\n",
      "Processing Bacteroides_intestinalis_61596 (26/200)\n",
      "Processing Bacteroides_massiliensis_44749 (27/200)\n",
      "Processing Bacteroides_ovatus_58035 (28/200)\n",
      "Processing Bacteroides_pectinophilus_61619 (29/200)\n",
      "Processing Bacteroides_plebeius_61623 (30/200)\n",
      "Processing Bacteroides_salyersiae_54873 (31/200)\n",
      "Processing Bacteroides_stercoris_56735 (32/200)\n",
      "Processing Bacteroides_thetaiotaomicron_56941 (33/200)\n",
      "Processing Bacteroides_uniformis_57318 (34/200)\n",
      "Processing Bacteroides_vulgatus_57955 (35/200)\n",
      "Processing Bacteroides_xylanisolvens_57185 (36/200)\n",
      "Processing Barnesiella_intestinihominis_62208 (37/200)\n",
      "Processing Bifidobacterium_adolescentis_56815 (38/200)\n",
      "Processing Bifidobacterium_angulatum_45832 (39/200)\n",
      "Processing Bifidobacterium_animalis_58116 (40/200)\n",
      "Processing Bifidobacterium_bifidum_55065 (41/200)\n",
      "Processing Bifidobacterium_catenulatum_58257 (42/200)\n",
      "Processing Bifidobacterium_longum_57796 (43/200)\n",
      "Processing Bifidobacterium_pseudocatenulatum_57754 (44/200)\n",
      "Processing Bilophila_wadsworthia_57364 (45/200)\n",
      "Processing Blautia_producta_56315 (46/200)\n",
      "Processing Blautia_wexlerae_56130 (47/200)\n",
      "Processing Burkholderiales_bacterium_56577 (48/200)\n",
      "Processing Butyricimonas_virosa_58742 (49/200)\n",
      "Processing Capnocytophaga_gingivalis_61780 (50/200)\n",
      "Processing Capnocytophaga_granulosa_57613 (51/200)\n",
      "Processing Capnocytophaga_sputigena_61779 (52/200)\n",
      "Processing Catenibacterium_mitsuokai_61547 (53/200)\n",
      "Processing Citrobacter_braakii_56022 (54/200)\n",
      "Processing Clostridiales_bacterium_56470 (55/200)\n",
      "Processing Clostridiales_bacterium_61057 (56/200)\n",
      "Processing Clostridium_bolteae_57158 (57/200)\n",
      "Processing Clostridium_citroniae_62210 (58/200)\n",
      "Processing Clostridium_sp_61482 (59/200)\n",
      "Processing Clostridium_spiroforme_61500 (60/200)\n",
      "Processing Collinsella_aerofaciens_61484 (61/200)\n",
      "Processing Collinsella_intestinalis_61689 (62/200)\n",
      "Processing Collinsella_sp_62205 (63/200)\n",
      "Processing Coprococcus_catus_62200 (64/200)\n",
      "Processing Coprococcus_comes_61587 (65/200)\n",
      "Processing Coprococcus_sp_62244 (66/200)\n",
      "Processing Coriobacteriaceae_bacterium_58375 (67/200)\n",
      "Processing Desulfovibrio_piger_61475 (68/200)\n",
      "Processing Dialister_invisus_61905 (69/200)\n",
      "Processing Dorea_formicigenerans_56346 (70/200)\n",
      "Processing Dorea_longicatena_59913 (71/200)\n",
      "Processing Dorea_longicatena_61473 (72/200)\n",
      "Processing Eggerthella_lenta_56463 (73/200)\n",
      "Processing Enterobacter_aerogenes_55704 (74/200)\n",
      "Processing Enterobacter_cloacae_55011 (75/200)\n",
      "Processing Enterobacter_cloacae_55612 (76/200)\n",
      "Processing Enterobacter_cloacae_57303 (77/200)\n",
      "Processing Enterobacter_cloacae_58148 (78/200)\n",
      "Processing Enterobacter_cloacae_58287 (79/200)\n",
      "Processing Enterococcus_durans_52177 (80/200)\n",
      "Processing Enterococcus_faecalis_56297 (81/200)\n",
      "Processing Enterococcus_faecium_56947 (82/200)\n",
      "Processing Enterococcus_hirae_55131 (83/200)\n",
      "Processing Escherichia_coli_58110 (84/200)\n",
      "Processing Eubacterium_biforme_61684 (85/200)\n",
      "Processing Eubacterium_cylindroides_56237 (86/200)\n",
      "Processing Eubacterium_eligens_61678 (87/200)\n",
      "Processing Eubacterium_hallii_61477 (88/200)\n",
      "Processing Eubacterium_limosum_60946 (89/200)\n",
      "Processing Eubacterium_ramulus_59802 (90/200)\n",
      "Processing Eubacterium_rectale_56927 (91/200)\n",
      "Processing Eubacterium_siraeum_57634 (92/200)\n",
      "Processing Eubacterium_sulci_62452 (93/200)\n",
      "Processing Eubacterium_ventriosum_61474 (94/200)\n",
      "Processing Faecalibacterium_cf_62236 (95/200)\n",
      "Processing Faecalibacterium_prausnitzii_57453 (96/200)\n",
      "Processing Faecalibacterium_prausnitzii_61481 (97/200)\n",
      "Processing Faecalibacterium_prausnitzii_62201 (98/200)\n",
      "Processing Fusobacterium_periodonticum_58002 (99/200)\n",
      "Processing Gemella_haemolysans_61762 (100/200)\n",
      "Processing Gemella_morbillorum_61817 (101/200)\n",
      "Processing Gemella_sanguinis_57791 (102/200)\n",
      "Processing Gordonibacter_pamelaeae_62044 (103/200)\n",
      "Processing Granulicatella_elegans_61945 (104/200)\n",
      "Processing Guyana_massiliensis_60772 (105/200)\n",
      "Processing Haemophilus_haemolyticus_58348 (106/200)\n",
      "Processing Haemophilus_parainfluenzae_62356 (107/200)\n",
      "Processing Haemophilus_pittmaniae_58383 (108/200)\n",
      "Processing Haemophilus_sputorum_53575 (109/200)\n",
      "Processing Kingella_oralis_61952 (110/200)\n",
      "Processing Klebsiella_oxytoca_54123 (111/200)\n",
      "Processing Klebsiella_oxytoca_56762 (112/200)\n",
      "Processing Klebsiella_oxytoca_57801 (113/200)\n",
      "Processing Klebsiella_pneumoniae_54688 (114/200)\n",
      "Processing Klebsiella_pneumoniae_54788 (115/200)\n",
      "Processing Klebsiella_pneumoniae_57337 (116/200)\n",
      "Processing Lachnospiraceae_bacterium_51870 (117/200)\n",
      "Processing Lactobacillus_acidophilus_51143 (118/200)\n",
      "Processing Lactobacillus_delbrueckii_56845 (119/200)\n",
      "Processing Lactobacillus_rhamnosus_57549 (120/200)\n",
      "Processing Lactobacillus_ruminis_57020 (121/200)\n",
      "Processing Lactococcus_garvieae_57061 (122/200)\n",
      "Processing Lautropia_mirabilis_62440 (123/200)\n",
      "Processing Leptotrichia_sp_60101 (124/200)\n",
      "Processing Megamonas_hypermegale_57114 (125/200)\n",
      "Processing Megasphaera_micronuciformis_62167 (126/200)\n",
      "Processing Megasphaera_sp_50408 (127/200)\n",
      "Processing Mitsuokella_jalaludinii_60453 (128/200)\n",
      "Processing Mitsuokella_multacida_61656 (129/200)\n",
      "Processing Neisseria_elongata_61756 (130/200)\n",
      "Processing Neisseria_sicca_58189 (131/200)\n",
      "Processing Neisseria_subflava_61760 (132/200)\n",
      "Processing Odoribacter_splanchnicus_62174 (133/200)\n",
      "Processing Oribacterium_sp_54059 (134/200)\n",
      "Processing Oscillibacter_sp_60799 (135/200)\n",
      "Processing Oscillospiraceae_bacterium_54867 (136/200)\n",
      "Processing Oxalobacter_formigenes_61802 (137/200)\n",
      "Processing Parabacteroides_distasonis_56985 (138/200)\n",
      "Processing Parabacteroides_goldsteinii_56831 (139/200)\n",
      "Processing Parabacteroides_johnsonii_55217 (140/200)\n",
      "Processing Parabacteroides_merdae_56972 (141/200)\n",
      "Processing Paraprevotella_clara_33712 (142/200)\n",
      "Processing Pediococcus_acidilactici_57243 (143/200)\n",
      "Processing Pediococcus_pentosaceus_55856 (144/200)\n",
      "Processing Phascolarctobacterium_sp_59817 (145/200)\n",
      "Processing Phascolarctobacterium_sp_59818 (146/200)\n",
      "Processing Porphyromonas_sp_57899 (147/200)\n",
      "Processing Prevotella_copri_61740 (148/200)\n",
      "Processing Prevotella_histicola_56864 (149/200)\n",
      "Processing Prevotella_intermedia_57633 (150/200)\n",
      "Processing Prevotella_melaninogenica_58075 (151/200)\n",
      "Processing Prevotella_nanceiensis_44721 (152/200)\n",
      "Processing Prevotella_oris_57058 (153/200)\n",
      "Processing Prevotella_pallens_62692 (154/200)\n",
      "Processing Prevotella_salivae_57216 (155/200)\n",
      "Processing Prevotella_shahii_59695 (156/200)\n",
      "Processing Prevotella_sp_58138 (157/200)\n",
      "Processing Prevotella_sp_58566 (158/200)\n",
      "Processing Prevotella_sp_61818 (159/200)\n",
      "Processing Propionibacterium_freudenreichii_62075 (160/200)\n",
      "Processing Raoultella_planticola_57814 (161/200)\n",
      "Processing Roseburia_hominis_61877 (162/200)\n",
      "Processing Roseburia_intestinalis_56239 (163/200)\n",
      "Processing Roseburia_inulinivorans_61943 (164/200)\n",
      "Processing Rothia_dentocariosa_57938 (165/200)\n",
      "Processing Ruminococcus_bicirculans_59300 (166/200)\n",
      "Processing Ruminococcus_bromii_62047 (167/200)\n",
      "Processing Ruminococcus_callidus_61479 (168/200)\n",
      "Processing Ruminococcus_gnavus_57638 (169/200)\n",
      "Processing Ruminococcus_lactaris_55568 (170/200)\n",
      "Processing Ruminococcus_obeum_61472 (171/200)\n",
      "Processing Ruminococcus_obeum_62046 (172/200)\n",
      "Processing Ruminococcus_sp_55468 (173/200)\n",
      "Processing Ruminococcus_torques_62045 (174/200)\n",
      "Processing Serratia_marcescens_59790 (175/200)\n",
      "Processing Streptococcus_anginosus_58223 (176/200)\n",
      "Processing Streptococcus_australis_62469 (177/200)\n",
      "Processing Streptococcus_infantarius_58294 (178/200)\n",
      "Processing Streptococcus_infantis_58320 (179/200)\n",
      "Processing Streptococcus_infantis_62384 (180/200)\n",
      "Processing Streptococcus_infantis_62471 (181/200)\n",
      "Processing Streptococcus_mitis_58556 (182/200)\n",
      "Processing Streptococcus_mitis_60474 (183/200)\n",
      "Processing Streptococcus_mitis_62363 (184/200)\n",
      "Processing Streptococcus_oralis_58560 (185/200)\n",
      "Processing Streptococcus_parasanguinis_58126 (186/200)\n",
      "Processing Streptococcus_pseudopneumoniae_58459 (187/200)\n",
      "Processing Streptococcus_salivarius_58037 (188/200)\n",
      "Processing Streptococcus_sp_54717 (189/200)\n",
      "Processing Streptococcus_sp_58554 (190/200)\n",
      "Processing Streptococcus_thermophilus_54772 (191/200)\n",
      "Processing Streptococcus_vestibularis_56030 (192/200)\n",
      "Processing Sutterella_wadsworthensis_56828 (193/200)\n",
      "Processing Sutterella_wadsworthensis_62218 (194/200)\n",
      "Processing Veillonella_atypica_58169 (195/200)\n",
      "Processing Veillonella_dispar_61763 (196/200)\n",
      "Processing Veillonella_parvula_57794 (197/200)\n",
      "Processing Veillonella_parvula_58184 (198/200)\n",
      "Processing Veillonella_sp_62404 (199/200)\n",
      "Processing Weissella_confusa_59158 (200/200)\n"
     ]
    }
   ],
   "source": [
    "metadata_map = parse_midas_data.parse_sample_metadata_map()\n",
    "samples_df = pd.DataFrame(columns = ['accession_id', 'species', 'haploid', 'subject_id'])\n",
    "\n",
    "accession_id_vec = []\n",
    "species_vec = []\n",
    "haploid_vec = []\n",
    "subject_id_vec = []\n",
    "\n",
    "counter = 0\n",
    "no_of_species = len(species_list)\n",
    "for species in species_list:\n",
    "    counter += 1\n",
    "    print(\"Processing \" + species + \" ({}/{})\".format(counter, no_of_species))\n",
    "    high_coverage_samples = diversity_utils.calculate_highcoverage_samples(species)\n",
    "    haploid_samples = diversity_utils.calculate_haploid_samples(species)\n",
    "    \n",
    "    haploid_boolean_vec = [True if sample in haploid_samples else False for sample in high_coverage_samples]\n",
    "    subject_id_vec_temp = [metadata_map[sample][0] for sample in high_coverage_samples]\n",
    "    \n",
    "    species_vec = species_vec + [species]*len(high_coverage_samples)\n",
    "    accession_id_vec = accession_id_vec + list(high_coverage_samples)\n",
    "    haploid_vec = haploid_vec + haploid_boolean_vec\n",
    "    subject_id_vec = subject_id_vec + subject_id_vec_temp\n",
    "\n",
    "samples_df['accession_id'] = accession_id_vec\n",
    "samples_df['species'] = species_vec\n",
    "samples_df['haploid'] = haploid_vec\n",
    "samples_df['subject_id'] = subject_id_vec\n",
    "\n",
    "    \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>accession_id</th>\n",
       "      <th>species</th>\n",
       "      <th>haploid</th>\n",
       "      <th>subject_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR18585109</td>\n",
       "      <td>Acidaminococcus_intestini_54097</td>\n",
       "      <td>True</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR18585184</td>\n",
       "      <td>Acidaminococcus_intestini_54097</td>\n",
       "      <td>False</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR18585205</td>\n",
       "      <td>Acidaminococcus_intestini_54097</td>\n",
       "      <td>True</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR18585207</td>\n",
       "      <td>Acidaminococcus_intestini_54097</td>\n",
       "      <td>True</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR18585211</td>\n",
       "      <td>Acidaminococcus_intestini_54097</td>\n",
       "      <td>True</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1540</th>\n",
       "      <td>SRR18585105</td>\n",
       "      <td>Veillonella_sp_62404</td>\n",
       "      <td>False</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1541</th>\n",
       "      <td>SRR18585106</td>\n",
       "      <td>Veillonella_sp_62404</td>\n",
       "      <td>False</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1542</th>\n",
       "      <td>SRR18585124</td>\n",
       "      <td>Veillonella_sp_62404</td>\n",
       "      <td>False</td>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1543</th>\n",
       "      <td>SRR18585145</td>\n",
       "      <td>Veillonella_sp_62404</td>\n",
       "      <td>False</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1544</th>\n",
       "      <td>SRR18585146</td>\n",
       "      <td>Veillonella_sp_62404</td>\n",
       "      <td>False</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1545 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     accession_id                          species  haploid subject_id\n",
       "0     SRR18585109  Acidaminococcus_intestini_54097     True         14\n",
       "1     SRR18585184  Acidaminococcus_intestini_54097    False         12\n",
       "2     SRR18585205  Acidaminococcus_intestini_54097     True          9\n",
       "3     SRR18585207  Acidaminococcus_intestini_54097     True          9\n",
       "4     SRR18585211  Acidaminococcus_intestini_54097     True          9\n",
       "...           ...                              ...      ...        ...\n",
       "1540  SRR18585105             Veillonella_sp_62404    False         14\n",
       "1541  SRR18585106             Veillonella_sp_62404    False         14\n",
       "1542  SRR18585124             Veillonella_sp_62404    False         13\n",
       "1543  SRR18585145             Veillonella_sp_62404    False          5\n",
       "1544  SRR18585146             Veillonella_sp_62404    False          5\n",
       "\n",
       "[1545 rows x 4 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plotting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_grouping = samples_df.groupby(['species', 'haploid']).size().reset_index(name='row_count')\n",
    "subject_grouping = samples_df.groupby(['subject_id', 'haploid']).size().reset_index(name='row_count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_grouping.groupby(by = ['species'])['row_count'].sum().reset_index().sort_values(by = 'row_count', ascending = False)#['species'].values\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_grouping[species_grouping.species == \"Bacteroides_vulgatus_57955\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(40, 8))\n",
    "sns.barplot(x = \"species\", y = \"row_count\", \n",
    "            data=species_grouping, \n",
    "            hue = \"haploid\", dodge = False, \n",
    "            ax = ax, \n",
    "            order = species_grouping.groupby(by = ['species'])['row_count'].sum().reset_index().sort_values(by = 'row_count', ascending = False)['species'].values\n",
    "\n",
    ")\n",
    "ax.set_xlabel(\"Species\", fontsize=20)\n",
    "ax.set_ylabel(\"Number of samples\", fontsize=20)\n",
    "ax.tick_params(axis='both', which='both', labelsize=20)\n",
    "#ax.set_title(\"QP Samples\", fontsize=30)\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=45, ha=\"right\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "figure_folder = \"%s%s\" % (config.analysis_directory, \"figures/\")\n",
    "out_file = \"%s%s\" % (figure_folder, \"QP_species_breakdown.png\")\n",
    "fig.savefig(out_file, dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "figure_folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(20, 8))\n",
    "sns.barplot(x = \"subject_id\", y = \"row_count\", \n",
    "            data=subject_grouping, \n",
    "            hue = \"haploid\", dodge = False, \n",
    "            order = [\"1\", \"2\", \"3\", \"4\", \"5\", \"6\", \"7\", \"8\", \"9\", \"10\", \"11\", \"12\", \"13\", \"14\", \"15\"],\n",
    "            ax = ax)\n",
    "ax.set_xlabel(\"Subject\", fontsize=20)\n",
    "ax.set_ylabel(\"Number of samples\", fontsize=20)\n",
    "ax.tick_params(axis='both', which='both', labelsize=20)\n",
    "ax.set_title(\"QP Samples\", fontsize=30)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "figure_folder = \"%s%s\" % (config.analysis_directory, \"figures/\")\n",
    "out_file = \"%s%s\" % (figure_folder, \"QP_subject_breakdown.png\")\n",
    "fig.savefig(out_file, dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(samples_df.accession_id.unique())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python3.8_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
