{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# paths\n",
    "## Metadata\n",
    "metadata_dir = \"/u/project/ngarud/michaelw/hsiao_data/tables/\"\n",
    "## sample report\n",
    "sample_report_path = \"%ssample_report.csv\" % (metadata_dir)\n",
    "## output path\n",
    "read_mapping_out_path = \"%ssample_report_Mv3_ReadMapping.csv\" % (metadata_dir)\n",
    "## MIDASv3 output directory\n",
    "MIDASv3_output_dir = \"/u/project/ngarud/Garud_lab/hsiao_data/MIDASv3_output/\"\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load sample report\n",
    "sample_report = pd.read_csv(sample_report_path, index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# accession list\n",
    "accession_list = [accession.split(\"_\")[0] for accession in sample_report.Accession]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# build mapped reads dictionary\n",
    "mapped_reads_dict = dict({})\n",
    "for accession in accession_list:\n",
    "    # summary paths\n",
    "    genes_summary_path = \"%s%s/genes/genes_summary.tsv\" % (MIDASv3_output_dir, accession)\n",
    "    snps_summary_path = \"%s%s/snps/snps_summary.tsv\" % (MIDASv3_output_dir, accession)\n",
    "\n",
    "    # load summary files\n",
    "    genes_summary = pd.read_csv(genes_summary_path, sep = \"\\t\")\n",
    "    snps_summary = pd.read_csv(snps_summary_path, sep = \"\\t\")\n",
    "\n",
    "    # calculate mapped reads\n",
    "    genes_mapped_reads = genes_summary.sum()['mapped_reads']\n",
    "    snps_mapped_reads = snps_summary.sum()['mapped_reads']\n",
    "\n",
    "    # add to dictionary\n",
    "    mapped_reads_dict[accession] = dict({\"genes\": genes_mapped_reads, \"snps\": snps_mapped_reads})\n",
    "\n",
    "\n",
    "    \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# annotate sample report file with mapped read counts\n",
    "sample_report['MIDASv3_Mapped_Reads_genes'] = sample_report.Accession.str.split(\"_\").str[0].apply(lambda x: mapped_reads_dict[x]['genes'])\n",
    "sample_report['MIDASv3_Mapped_Reads_snps'] = sample_report.Accession.str.split(\"_\").str[0].apply(lambda x: mapped_reads_dict[x]['snps'])\n",
    "sample_report['MIDASv3_PercentMapped_genes'] = sample_report['MIDASv3_Mapped_Reads_genes']/sample_report['Total_coverage']\n",
    "sample_report['MIDASv3_PercentMapped_snps'] = sample_report['MIDASv3_Mapped_Reads_snps']/sample_report['Total_coverage']\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Saving\n",
    "sample_report.to_csv(read_mapping_out_path, index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "## TEST\n",
    "accession = \"35-2mo-3-hTau-B2-Female\"\n",
    "# summary paths\n",
    "genes_summary_path = \"%s%s/genes/genes_summary.tsv\" % (MIDASv3_output_dir, accession)\n",
    "snps_summary_path = \"%s%s/snps/snps_summary.tsv\" % (MIDASv3_output_dir, accession)\n",
    "\n",
    "# load summary files\n",
    "genes_summary = pd.read_csv(genes_summary_path, sep = \"\\t\")\n",
    "snps_summary = pd.read_csv(snps_summary_path, sep = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>species_id</th>\n",
       "      <th>pangenome_size</th>\n",
       "      <th>covered_genes</th>\n",
       "      <th>fraction_covered</th>\n",
       "      <th>aligned_reads</th>\n",
       "      <th>mapped_reads</th>\n",
       "      <th>mean_depth</th>\n",
       "      <th>marker_depth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>142728</td>\n",
       "      <td>3961</td>\n",
       "      <td>3606</td>\n",
       "      <td>0.910</td>\n",
       "      <td>207824</td>\n",
       "      <td>177516</td>\n",
       "      <td>6.886</td>\n",
       "      <td>6.102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>141329</td>\n",
       "      <td>1687</td>\n",
       "      <td>706</td>\n",
       "      <td>0.418</td>\n",
       "      <td>83546</td>\n",
       "      <td>73888</td>\n",
       "      <td>16.533</td>\n",
       "      <td>4.883</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>141274</td>\n",
       "      <td>811</td>\n",
       "      <td>691</td>\n",
       "      <td>0.852</td>\n",
       "      <td>120723</td>\n",
       "      <td>107997</td>\n",
       "      <td>20.426</td>\n",
       "      <td>18.437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>138779</td>\n",
       "      <td>5176</td>\n",
       "      <td>4550</td>\n",
       "      <td>0.879</td>\n",
       "      <td>693015</td>\n",
       "      <td>604124</td>\n",
       "      <td>18.777</td>\n",
       "      <td>15.869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>137981</td>\n",
       "      <td>3491</td>\n",
       "      <td>3080</td>\n",
       "      <td>0.882</td>\n",
       "      <td>164046</td>\n",
       "      <td>135182</td>\n",
       "      <td>6.411</td>\n",
       "      <td>3.285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>137434</td>\n",
       "      <td>2035</td>\n",
       "      <td>1897</td>\n",
       "      <td>0.932</td>\n",
       "      <td>185924</td>\n",
       "      <td>154037</td>\n",
       "      <td>10.265</td>\n",
       "      <td>9.306</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>132831</td>\n",
       "      <td>3971</td>\n",
       "      <td>3439</td>\n",
       "      <td>0.866</td>\n",
       "      <td>619029</td>\n",
       "      <td>557685</td>\n",
       "      <td>21.129</td>\n",
       "      <td>19.156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>130685</td>\n",
       "      <td>2863</td>\n",
       "      <td>2802</td>\n",
       "      <td>0.979</td>\n",
       "      <td>153947</td>\n",
       "      <td>110762</td>\n",
       "      <td>5.875</td>\n",
       "      <td>4.514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>129214</td>\n",
       "      <td>4704</td>\n",
       "      <td>3890</td>\n",
       "      <td>0.827</td>\n",
       "      <td>182932</td>\n",
       "      <td>153125</td>\n",
       "      <td>5.895</td>\n",
       "      <td>5.107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>128931</td>\n",
       "      <td>2226</td>\n",
       "      <td>2052</td>\n",
       "      <td>0.922</td>\n",
       "      <td>465422</td>\n",
       "      <td>424369</td>\n",
       "      <td>30.023</td>\n",
       "      <td>31.121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>128185</td>\n",
       "      <td>1902</td>\n",
       "      <td>1786</td>\n",
       "      <td>0.939</td>\n",
       "      <td>13113381</td>\n",
       "      <td>12254165</td>\n",
       "      <td>824.550</td>\n",
       "      <td>758.347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>121662</td>\n",
       "      <td>3039</td>\n",
       "      <td>2369</td>\n",
       "      <td>0.780</td>\n",
       "      <td>104206</td>\n",
       "      <td>83845</td>\n",
       "      <td>5.224</td>\n",
       "      <td>4.722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>121364</td>\n",
       "      <td>1372</td>\n",
       "      <td>1241</td>\n",
       "      <td>0.905</td>\n",
       "      <td>68977</td>\n",
       "      <td>63391</td>\n",
       "      <td>7.575</td>\n",
       "      <td>8.478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>118079</td>\n",
       "      <td>1245</td>\n",
       "      <td>1126</td>\n",
       "      <td>0.904</td>\n",
       "      <td>61906</td>\n",
       "      <td>57108</td>\n",
       "      <td>7.328</td>\n",
       "      <td>5.519</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>114332</td>\n",
       "      <td>2241</td>\n",
       "      <td>1761</td>\n",
       "      <td>0.786</td>\n",
       "      <td>90427</td>\n",
       "      <td>77372</td>\n",
       "      <td>5.156</td>\n",
       "      <td>3.712</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>112332</td>\n",
       "      <td>7301</td>\n",
       "      <td>5981</td>\n",
       "      <td>0.819</td>\n",
       "      <td>1214921</td>\n",
       "      <td>1055843</td>\n",
       "      <td>23.979</td>\n",
       "      <td>25.895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>110884</td>\n",
       "      <td>4859</td>\n",
       "      <td>4091</td>\n",
       "      <td>0.842</td>\n",
       "      <td>357150</td>\n",
       "      <td>295579</td>\n",
       "      <td>10.140</td>\n",
       "      <td>8.410</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>109339</td>\n",
       "      <td>2245</td>\n",
       "      <td>2096</td>\n",
       "      <td>0.934</td>\n",
       "      <td>975651</td>\n",
       "      <td>894954</td>\n",
       "      <td>63.742</td>\n",
       "      <td>57.813</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>108721</td>\n",
       "      <td>9496</td>\n",
       "      <td>5841</td>\n",
       "      <td>0.615</td>\n",
       "      <td>276081</td>\n",
       "      <td>218475</td>\n",
       "      <td>5.859</td>\n",
       "      <td>2.601</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>107220</td>\n",
       "      <td>4786</td>\n",
       "      <td>3595</td>\n",
       "      <td>0.751</td>\n",
       "      <td>301607</td>\n",
       "      <td>247107</td>\n",
       "      <td>9.192</td>\n",
       "      <td>6.751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>103955</td>\n",
       "      <td>3778</td>\n",
       "      <td>2854</td>\n",
       "      <td>0.755</td>\n",
       "      <td>11665209</td>\n",
       "      <td>10882202</td>\n",
       "      <td>472.045</td>\n",
       "      <td>584.379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>103753</td>\n",
       "      <td>3492</td>\n",
       "      <td>2484</td>\n",
       "      <td>0.711</td>\n",
       "      <td>479679</td>\n",
       "      <td>358545</td>\n",
       "      <td>21.150</td>\n",
       "      <td>18.187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>101347</td>\n",
       "      <td>7694</td>\n",
       "      <td>5065</td>\n",
       "      <td>0.658</td>\n",
       "      <td>23756043</td>\n",
       "      <td>21998632</td>\n",
       "      <td>546.648</td>\n",
       "      <td>865.091</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>101186</td>\n",
       "      <td>6773</td>\n",
       "      <td>4653</td>\n",
       "      <td>0.687</td>\n",
       "      <td>5111357</td>\n",
       "      <td>4462311</td>\n",
       "      <td>143.019</td>\n",
       "      <td>126.030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>101047</td>\n",
       "      <td>8450</td>\n",
       "      <td>5657</td>\n",
       "      <td>0.669</td>\n",
       "      <td>7067838</td>\n",
       "      <td>6265777</td>\n",
       "      <td>142.160</td>\n",
       "      <td>190.003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>100520</td>\n",
       "      <td>6016</td>\n",
       "      <td>3670</td>\n",
       "      <td>0.610</td>\n",
       "      <td>5834239</td>\n",
       "      <td>5367211</td>\n",
       "      <td>203.338</td>\n",
       "      <td>331.125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>100304</td>\n",
       "      <td>10842</td>\n",
       "      <td>6959</td>\n",
       "      <td>0.642</td>\n",
       "      <td>44111541</td>\n",
       "      <td>40804966</td>\n",
       "      <td>823.282</td>\n",
       "      <td>2423.151</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>100224</td>\n",
       "      <td>24052</td>\n",
       "      <td>12442</td>\n",
       "      <td>0.517</td>\n",
       "      <td>4967007</td>\n",
       "      <td>4609514</td>\n",
       "      <td>46.092</td>\n",
       "      <td>306.703</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    species_id  pangenome_size  covered_genes  fraction_covered  \\\n",
       "0       142728            3961           3606             0.910   \n",
       "1       141329            1687            706             0.418   \n",
       "2       141274             811            691             0.852   \n",
       "3       138779            5176           4550             0.879   \n",
       "4       137981            3491           3080             0.882   \n",
       "5       137434            2035           1897             0.932   \n",
       "6       132831            3971           3439             0.866   \n",
       "7       130685            2863           2802             0.979   \n",
       "8       129214            4704           3890             0.827   \n",
       "9       128931            2226           2052             0.922   \n",
       "10      128185            1902           1786             0.939   \n",
       "11      121662            3039           2369             0.780   \n",
       "12      121364            1372           1241             0.905   \n",
       "13      118079            1245           1126             0.904   \n",
       "14      114332            2241           1761             0.786   \n",
       "15      112332            7301           5981             0.819   \n",
       "16      110884            4859           4091             0.842   \n",
       "17      109339            2245           2096             0.934   \n",
       "18      108721            9496           5841             0.615   \n",
       "19      107220            4786           3595             0.751   \n",
       "20      103955            3778           2854             0.755   \n",
       "21      103753            3492           2484             0.711   \n",
       "22      101347            7694           5065             0.658   \n",
       "23      101186            6773           4653             0.687   \n",
       "24      101047            8450           5657             0.669   \n",
       "25      100520            6016           3670             0.610   \n",
       "26      100304           10842           6959             0.642   \n",
       "27      100224           24052          12442             0.517   \n",
       "\n",
       "    aligned_reads  mapped_reads  mean_depth  marker_depth  \n",
       "0          207824        177516       6.886         6.102  \n",
       "1           83546         73888      16.533         4.883  \n",
       "2          120723        107997      20.426        18.437  \n",
       "3          693015        604124      18.777        15.869  \n",
       "4          164046        135182       6.411         3.285  \n",
       "5          185924        154037      10.265         9.306  \n",
       "6          619029        557685      21.129        19.156  \n",
       "7          153947        110762       5.875         4.514  \n",
       "8          182932        153125       5.895         5.107  \n",
       "9          465422        424369      30.023        31.121  \n",
       "10       13113381      12254165     824.550       758.347  \n",
       "11         104206         83845       5.224         4.722  \n",
       "12          68977         63391       7.575         8.478  \n",
       "13          61906         57108       7.328         5.519  \n",
       "14          90427         77372       5.156         3.712  \n",
       "15        1214921       1055843      23.979        25.895  \n",
       "16         357150        295579      10.140         8.410  \n",
       "17         975651        894954      63.742        57.813  \n",
       "18         276081        218475       5.859         2.601  \n",
       "19         301607        247107       9.192         6.751  \n",
       "20       11665209      10882202     472.045       584.379  \n",
       "21         479679        358545      21.150        18.187  \n",
       "22       23756043      21998632     546.648       865.091  \n",
       "23        5111357       4462311     143.019       126.030  \n",
       "24        7067838       6265777     142.160       190.003  \n",
       "25        5834239       5367211     203.338       331.125  \n",
       "26       44111541      40804966     823.282      2423.151  \n",
       "27        4967007       4609514      46.092       306.703  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_summary"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py38",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
