{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, \"/u/home/m/michaelw/project-ngarud/microbiome_evolution/microbiome_evolution_MOUSE/\")\n",
    "\n",
    "import numpy\n",
    "import sys\n",
    "import config\n",
    "import gzip\n",
    "import os.path\n",
    "import os\n",
    "import midas_db_utils\n",
    "\n",
    "#My additions\n",
    "import core_gene_utils\n",
    "from core_gene_utils import *\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Directories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "core_genes_directory = (\"%score_genes/\" % config.data_directory)\n",
    "external_core_genes_directory = (\"%score_genes/external/\" % config.data_directory)\n",
    "\n",
    "default_external_shared_gene_filename = (external_core_genes_directory+\"shared_genes.txt.gz\")\n",
    "default_external_core_gene_filename = (external_core_genes_directory+\"core_genes.txt.gz\")\n",
    "default_external_stringent_core_gene_filename = (external_core_genes_directory+\"core_genes_stringent.txt.gz\")\n",
    "default_external_gene_freq_template = (external_core_genes_directory+\"%s_gene_freqs.txt.gz\")\n",
    "\n",
    "default_shared_gene_filename = (core_genes_directory+\"shared_genes_HMP.txt.gz\")\n",
    "default_core_gene_filename = (core_genes_directory+\"core_genes.txt.gz\")\n",
    "default_stringent_core_gene_filename = (core_genes_directory+\"core_genes_stringent.txt.gz\")\n",
    "default_gene_freq_template = (core_genes_directory+\"%s_gene_freqs.txt.gz\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MAIN"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Importing another script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import parse_midas_data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Making core gene directory if needed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.system('mkdir -p %s' % core_genes_directory)\n",
    "os.system('mkdir -p %s' % external_core_genes_directory)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Parsing good species"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The requirements:\n",
    "- good_species_min_coverage = 10\n",
    "- good_species_min_prevalence = 10\n",
    "\n",
    "Using the function:\n",
    "\n",
    "> species_coverage_matrix, samples, species = parse_midas_data.parse_global_marker_gene_coverages()\n",
    "\n",
    "it generates an list made up of arrays equal to the number of SNPs, with each array being the length of the number of samples. \"samples\" and \"species\" provide a key for this marker gene coverage list.\n",
    "\n",
    "parse_good_species_list() then iterates through each species generated by parse_global_marker_gene_coverages(), and checks:\n",
    "- the number of samples with > the min marker gene coverage is > the prevalence threshold\n",
    "\n",
    "It appends all species that pass to \"good_species_list,\" which here is assigned to **pangenome_species**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "pangenome_species = parse_midas_data.parse_good_species_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Blautia_producta_56315',\n",
       " 'Akkermansia_muciniphila_55290',\n",
       " 'Roseburia_intestinalis_56239',\n",
       " 'Bacteroides_cellulosilyticus_58046',\n",
       " 'Escherichia_coli_58110',\n",
       " 'Adlercreutzia_equolifaciens_60310',\n",
       " 'Enterococcus_faecium_56947',\n",
       " 'Parabacteroides_distasonis_56985',\n",
       " 'Bacteroides_ovatus_58035',\n",
       " 'Blautia_wexlerae_56130',\n",
       " 'Ruminococcus_sp_58571',\n",
       " 'Burkholderiales_bacterium_56577',\n",
       " 'Lactobacillus_acidophilus_51143',\n",
       " 'Coprococcus_sp_62244',\n",
       " 'Bacteroides_vulgatus_57955',\n",
       " 'Bifidobacterium_animalis_58116',\n",
       " 'Anaerostipes_hadrus_55206',\n",
       " 'Bacteroides_uniformis_57318',\n",
       " 'Clostridiales_bacterium_52743',\n",
       " 'Ruminococcus_bicirculans_59300',\n",
       " 'Clostridium_clostridioforme_51842',\n",
       " 'Clostridiales_bacterium_61057',\n",
       " 'Faecalibacterium_prausnitzii_61481',\n",
       " 'Sutterella_wadsworthensis_56828',\n",
       " 'Eubacterium_hallii_61477',\n",
       " 'Ruminococcus_bromii_62047',\n",
       " 'Lactococcus_lactis_57073',\n",
       " 'Lachnospiraceae_bacterium_56833',\n",
       " 'Bacteroides_thetaiotaomicron_56941',\n",
       " 'Bacteroides_massiliensis_44749',\n",
       " 'Ruminococcus_gauvreauii_59033',\n",
       " 'Ruminococcus_obeum_61472',\n",
       " 'Clostridium_symbiosum_54029',\n",
       " 'Guyana_massiliensis_60772',\n",
       " 'Alistipes_shahii_62199',\n",
       " 'Alistipes_putredinis_61533',\n",
       " 'Barnesiella_intestinihominis_62208',\n",
       " 'Alistipes_finegoldii_56071',\n",
       " 'Bacteroides_fragilis_54507',\n",
       " 'Lachnospiraceae_bacterium_51870',\n",
       " 'Clostridiales_bacterium_52957']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pangenome_species"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Setting coverage minimums"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "cmin (config.core_genome_min_copynum) and cmax (config.core_genome_max_copynum) set the thresholds beyond which gene are not considered as part of the core genome.\n",
    "\n",
    "I'm not sure what shared shared_cminâ€”hopefully that will become clear!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "min_good_fraction (config.core_genome_min_prevalence) is the fraction of samples a gene needs to be in to be considered \"core.\" This is **0.9**.\n",
    "\n",
    "The minimum coverage for a gene to be included in the core genome is 5. I'm not sure if this is an average, or per sample."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "cmin = config.core_genome_min_copynum\n",
    "cmax = config.core_genome_max_copynum  \n",
    "shared_cmin = config.shared_genome_min_copynum"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "cmin is 0.3\n",
      "cmax is 3\n",
      "shared_cmin is 3\n"
     ]
    }
   ],
   "source": [
    "print \"cmin is %s\" % (cmin)\n",
    "print \"cmax is %s\" % (cmax)\n",
    "print \"shared_cmin is %s\" % (shared_cmin)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "min_good_fraction = config.core_genome_min_prevalence\n",
    "min_coverage = 5 # (for assessing core genome, we'll use a lower coverage value than when we look at real changes)\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "min_good_fraction is 0.9\n"
     ]
    }
   ],
   "source": [
    "print \"min_good_fraction is %s\" % (min_good_fraction)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Opening the files I'm writing to"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_filename = default_core_gene_filename\n",
    "#output_file = gzip.GzipFile(output_filename,\"w\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "stringent_output_filename = default_stringent_core_gene_filename\n",
    "#stringent_output_file = gzip.GzipFile(stringent_output_filename,\"w\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#shared_output_file = gzip.GzipFile(default_shared_gene_filename,\"w\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### the MAIN for loop!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This loops through each species in **pangenome_species**. I will use  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "!ls /u/project/ngarud/Garud_lab/HumanizedMouse/HumanizedMouse_Batch2/merged_data/snps/ | grep \"coli\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_name = \"Anaerostipes_hadrus_55206\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load reference genes\n",
    "sys.stderr.write(\"Loading genes on reference genome..\\n\")\n",
    "reference_genes = midas_db_utils.load_reference_genes(species_name) #Self explanatory\n",
    "sys.stderr.write(\"Done!\\n\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load reference genes\n",
    "sys.stderr.write(\"Loading shared genes from midas db..\\n\")\n",
    "midas_shared_genes = midas_db_utils.parse_midas_shared_genes(species_name)             \n",
    "sys.stderr.write(\"Done!\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here's what `midas_db_utils.parse_midas_shared_genes(species_name)` doing:\n",
    "1. It initializes `midas_shared_genes` as an empty set\n",
    "2. It loads `centroid_gene_map`\n",
    "    - This is taking the `gene_info.txt.gz` file for a given species (e.g., `/u/project/ngarud/Garud_lab/midas_db_v1.2/pan_genomes/Bacteroides_uniformis_57318/gene_info.txt.gz`), and creating a dictionary (**centroid_gene_map**).\n",
    "    - This dictionary uses `centroid_95` as the dictionary key, and either uses:\n",
    "        - `gene_id` as the item if the `gene_id` is in the reference genome and the `centroid_95` is not in the reference genome. Otherwise, it uses `centroid_95` as the key.\n",
    "3. It loads `cross_species_centroids.txt.gz` (`'/u/project/ngarud/Garud_lab/midas_db_v1.2/cross_species_centroids.txt.gz'`), which represents gene centroids observed across species boundaries.\n",
    "4. For each line of the file, it\n",
    "    - initializes `midas_shared_genes` as an empty set.\n",
    "    - It assigns the first column to `big_centroid`, and appends `big_centroid` to the `midas_shared_genes` set.\n",
    "    - it assigns the second column (which could hold multiple genes) to a vector called `other_centroids`\n",
    "    - it loops through each `other_centroid`, and appends these centroids to `midas_shared_genes` if they are a key in the `centroid_gene_map`\n",
    "5. it outputs `midas_shared_genes`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bad_pangenome_data = False #it starts as false, but may be set to true later"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load gene coverage information for species_name\n",
    "sys.stderr.write(\"Loading pangenome data for %s...\\n\" % species_name)\n",
    "gene_samples, gene_names, gene_presence_matrix, gene_depth_matrix, marker_coverages, gene_reads_matrix = parse_midas_data.parse_pangenome_data(species_name)\n",
    "sys.stderr.write(\"Done!\\n\")  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"245018.3.peg.2048\" in gene_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(marker_coverages)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Where:\n",
    "- `gene_samples` is a list of samples for which genes were calculated in the MIDAS pipeline (e.g., for *B. uniformis*, their are 33 samples for which gene information was calculated)\n",
    "- `gene_names` is a list of gene names (e.g., there are 8914 for *B. uniformis*)\n",
    "- `gene_presence_matrix` is a matrix of dimensions [# of genes, # of samples] that contains a boolean outcome for presence/absence (e.g., for *B. uniformis*, it has dimensions [8914, 33])\n",
    "- `gene_depth_matrix` has the same dimensions as `gene_presence_matrix`, but is filled with average gene depth values.\n",
    "- `marker_coverages` is a vector of marker coverages for each sample (e.g., for *B. uniformis*, it has 33 values for all 33 samples).\n",
    "- `gene_reads_matrix` has the same dimensions as `gene_presence_matrix`, but is filled with the number of reads mapping to each gene (in other words, it's not normalized by gene length). \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(marker_coverages)==0:\n",
    "    bad_pangenome_data = True\n",
    "else:        \n",
    "\n",
    "    high_coverage_idxs = (marker_coverages>=min_coverage) #i.e., selecting all samples with average coverage over 5\n",
    "\n",
    "    if high_coverage_idxs.sum() < 0.5: #This is essentally asking if there are 0 high coverage idxs\n",
    "        bad_pangenome_data = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if bad_pangenome_data:\n",
    "    # Just use reference genes\n",
    "    sys.stderr.write(\"Bad pangenome data for %s!\\n\" % species_name)\n",
    "    shared_gene_names = sorted(midas_shared_genes)\n",
    "    core_gene_names = sorted(reference_genes - midas_shared_genes)\n",
    "    stringent_gene_names = sorted(reference_genes - midas_shared_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "if bad_pangenome_data == True, then the core genes that are written out are based on the reference genome. Specifally:\n",
    "- shared genes is those that were calculated in this script\n",
    "- core genes are those in the reference that *aren't* present in the shared genes vector\n",
    "    - stringent_gene_names is the same thing\n",
    "    \n",
    "    \n",
    "The rest of this exploration script will be for if the bad_pangenome_data == False."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_names = numpy.array(gene_names) #created by parse_midas_data.parse_pangenome_data()\n",
    "gene_samples = gene_samples[high_coverage_idxs] #gene_samples created from parse_midas_data.parse_pangenome_data(). We're subsetting to only include \n",
    "marker_coverages = marker_coverages[high_coverage_idxs] #marker_coverages created from parse_midas_data.parse_pangenome_data(). We're subsetting to only include high coverage samples\n",
    "gene_depth_matrix = gene_depth_matrix[:,high_coverage_idxs] #gene_depth_matrix created from parse_midas_data.parse_pangenome_data(). We're subsetting to only include high coverage samples\n",
    "gene_copynum_matrix = gene_depth_matrix*1.0/(marker_coverages+(marker_coverages==0)) #gene_depth_matrix and marker coverage created by parse_midas_data.parse_pangenome_data(). Here, we convert them into copy number using universal single copy marker genes.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "good_sample_idxs = get_good_pangenome_samples(marker_coverages, gene_copynum_matrix, species_name)\n",
    "bad_sample_idxs = numpy.logical_not(good_sample_idxs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What is `get_good_pangenome_samples` doing?\n",
    "1. calculates the number of reference genes (using `parse_midas_data.load_reference_genes(species_name)`)\n",
    "2. counts the number of present genes in each sample by calculating how many elements in `gene_copynum_matrix` are over cmin (0.3) and summing the resulting columns.\n",
    "3. counts the number of genes in each sample over our high gene threshold by calculating how many elements in `gene_copynum_matrix` are over cmax (3) and summing the resulting columns. \n",
    "4. What is `(num_present_genes>0.3*num_reference_genes)*(num_high_genes<0.3*num_present_genes)` doing?\n",
    "    - `(num_present_genes>0.3*num_reference_genes)` identifies the samples that have > $0.3 \\times \\text{the number of reference genes}$\n",
    "    - `(num_high_genes<0.3*num_present_genes)` identifies the samples with < $0.3 \\times \\text{the number of genes present (in each sample)}$\n",
    "    - multiplying the two together *only* produces a true in the output vector if there is a true at the same position in the two input vectors\n",
    "    - ***essentially, this selects for samples that have both at least 30% of the reference genome present, and with no more than 30% of genes have copy number over 3.***\n",
    "\n",
    "What is `bad_sample_idxs`?\n",
    "- this just reverses the boolean operator in the `good_sample_idxs` vector"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_samples = gene_samples[good_sample_idxs] #This subsets gene samples by those that pass the thresholds specified above\n",
    "marker_coverages = marker_coverages[good_sample_idxs] #This does the same for our marker_coverages\n",
    "gene_copynum_matrix = gene_copynum_matrix[:,good_sample_idxs] #This does the same for the columns of our gene_copynum_matrix\n",
    "\n",
    "reference_gene_idxs = numpy.array([gene_name in reference_genes for gene_name in gene_names]) #This selects the gene names that are in the reference genome (boolean)\n",
    "\n",
    "midas_shared_idxs = numpy.array([gene_name in midas_shared_genes for gene_name in gene_names]) #This selects the gene names of genes shared across species (boolean)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "metagenome_shared_idxs = ((gene_copynum_matrix>shared_cmin).sum(axis=1)>0.5) #This is summing the number of samples in which a gene has over\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What is `metagenome_shared_idxs`?\n",
    "- This is a boolean vector the length of the number of genes\n",
    "- each element indicates whether that gene has a copy number >3 in *any* sample\n",
    "\n",
    "If a gene has CN>3 in any sample, this provides evidence that its shared between species (hence the name)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Now union with those we identified from midas db\n",
    "shared_idxs = numpy.logical_or(metagenome_shared_idxs, midas_shared_idxs) #if at least one is true, then output is true\n",
    "non_shared_idxs = numpy.logical_not(shared_idxs) #flips the booleans in shared_idxs\n",
    "shared_gene_names = gene_names[shared_idxs] #slices gene_names to only include those that are shared across species\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# calculating good genes\n",
    "good_idxs = (((gene_copynum_matrix>=cmin)*(gene_copynum_matrix<=cmax)).sum(axis=1)*1.0/len(marker_coverages) >= min_good_fraction)\n",
    "core_gene_idxs = good_idxs*reference_gene_idxs*non_shared_idxs\n",
    "core_gene_names = gene_names[core_gene_idxs]\n",
    "\n",
    "num_metagenome_and_midas = numpy.logical_and(midas_shared_idxs, metagenome_shared_idxs).sum()\n",
    "num_metagenome_only = numpy.logical_and(metagenome_shared_idxs, numpy.logical_not(midas_shared_idxs)).sum()\n",
    "num_midas_only = numpy.logical_and(midas_shared_idxs, numpy.logical_not(metagenome_shared_idxs)).sum()\n",
    "num_metagenome_or_midas = shared_idxs.sum()\n",
    "num_remaining = non_shared_idxs.sum()\n",
    "num_reference_remaining = (non_shared_idxs*reference_gene_idxs).sum()\n",
    "num_core = core_gene_idxs.sum()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**good_idxs:** \n",
    "1. For each gene, the number of samples with 0.3<GCN<3 is divided by the number of good samples\n",
    "2. If it's greater than `min_good_fraction` (i.e., 0.9), it's 1, if not, it's 0. \n",
    "3. output: a vector of 1s and 0s corresponding to all genes\n",
    "**core_gene_idxs:**\n",
    "1. multiplies the boolean vectors describing good indices, reference gene indices, and non-shared indices\n",
    "2. Only genes that are true in all three are true in the output\n",
    "**core_gene_names:** using `core_gene_idxs` to retrieve core gene names\n",
    "\n",
    "**STATS:**\n",
    "- `num_metagenome_and_midas`: number of genes that are shared as determined by **both** midas_db_utils and from the metagenomic data itself\n",
    "- `num_metagenome_only`: number of genes that are shared as calculated by metagenomic data only (that is, excluding those that are calculated from midas_db_utils)\n",
    "- `num_midas_only`: umber of genes that are shared as calculated by midas_db_utils only (that is, excluding those calculated directly from the data)\n",
    "- `num_metagenome_or_midas`: total number of shared genes, as calculated by **either** midas_db_utils and from the metagenomic data itself\n",
    "- `num_remaining`: number of non-shared genes\n",
    "- `num_reference_remaining`: number of non-shared reference genes\n",
    "- `num_core`: number of core genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print \"%s %d %d %d %d %d %d %d\" % (species_name, num_metagenome_and_midas, num_metagenome_only, num_midas_only, num_metagenome_or_midas, num_remaining, num_reference_remaining, num_core)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Measure frequencies and output them\n",
    "gene_prevalence_numerators = ((gene_copynum_matrix>=cmin)*(gene_copynum_matrix<=cmax)).sum(axis=1)\n",
    "gene_prevalence_denominators = ((gene_copynum_matrix<=cmax).sum(axis=1))\n",
    "\n",
    "good_prevalence_idxs = (gene_prevalence_numerators>0.5)*(gene_prevalence_denominators>0.5)*non_shared_idxs\n",
    "\n",
    "gene_prevalence_names = gene_names[good_prevalence_idxs]\n",
    "gene_prevalences = gene_prevalence_numerators[good_prevalence_idxs]*1.0/gene_prevalence_denominators[good_prevalence_idxs]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "`gene_prevalence_numerators`: For each gene, the number of samples in which the gene with 0.3<CN<3 (i.e., present)\n",
    "`gene_prevalence_denominators`: For each gene, the number of samples in which CN<3 (i.e., not shared, possibly present, possibly absent)\n",
    "`good_prevalence_idxs`: this is a boolen vector for all genes. The value for a gene is true if it's present in at least one sample, has a CN < 3 in at least one sample (kinda redundant), and not shared (also kinda redundant). \n",
    "`gene_prevalence_names`: names of non-shared genes that are prevalent in at least on sample\n",
    "`gene_prevalences`: The actual prevalence of each gene (only considering non-shared genes that are prevalent in at least on sample)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MW: Addition to add shared genes we detect in our data..."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Step 1: load in shared genes file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "HMP_shared_genes = parse_shared_genes(species_name, default_shared_gene_filename, default_external_shared_gene_filename)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "parse_shared_genes(species_name, default_shared_gene_filename, default_external_shared_gene_filename)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "HMP_shared_genes.update(set(shared_gene_names))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "numpy.array(list(HMP_shared_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_1 = set([0,1,2,3,4])\n",
    "set_2  = set([1,2,3,4,5])\n",
    "\n",
    "set_2.update(set_1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set_1 = set([0,1,2])\n",
    "set_2 = set([3])\n",
    "set_1.update(set_2)\n",
    "set_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(set_1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "[gid for gids_in_ena if gid not in sample_text_gid]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "HMP_shared_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'1095771.3.peg.969' in shared_gene_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "HMP_core_genes = parse_core_genes(species_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "HMP_core_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(set(shared_gene_names).intersection(HMP_shared_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(HMP_shared_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len(shared_gene_names)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python27_env",
   "language": "python",
   "name": "python27_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
