{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 1: is `calculate_snp_prevalences.parse_population_freqs(species_name, use_HMP_freqs=True)` loading HMP snp prevalences?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, \"/u/home/m/michaelw/project-ngarud/microbiome_evolution/microbiome_evolution_MOUSE/\")\n",
    "import numpy\n",
    "import bz2\n",
    "import gzip\n",
    "import config\n",
    "import os.path\n",
    "\n",
    "intermediate_filename_template = config.data_directory+\"snp_prevalences/%s.txt.gz\"\n",
    "intermediate_filename_template_HMP = config.HMP_data_directory+\"snp_prevalences/%s.txt.gz\" #MW added this \n",
    "\n",
    "def parse_population_freqs(desired_species_name, polarize_by_consensus=False, use_HMP_freqs=False):\n",
    "    \n",
    "    if use_HMP_freqs:                                                                           #MW added this\n",
    "        intermediate_filename = intermediate_filename_template_HMP % desired_species_name       #if statement.\n",
    "    else:                                                                                       #If statement will be\n",
    "        intermediate_filename = intermediate_filename_template % desired_species_name           #true in calculate_within_person_sfs.py\n",
    "      \n",
    "    population_freqs = {}\n",
    "    \n",
    "    if not os.path.isfile(intermediate_filename):\n",
    "        return population_freqs\n",
    "\n",
    "    file = gzip.GzipFile(intermediate_filename,\"r\")\n",
    "    file.readline()\n",
    "    for line in file:\n",
    "        items = line.split(\",\")\n",
    "        contig = items[0]\n",
    "        location = long(items[1])\n",
    "        population_freq = float(items[2])\n",
    "        snp_freq = float(items[3])\n",
    "        \n",
    "        if polarize_by_consensus:\n",
    "            if population_freq > 0.5:\n",
    "                population_freq = 1-population_freq\n",
    "        \n",
    "        if population_freq==0:\n",
    "            pass\n",
    "        else:\n",
    "            population_freqs[(contig,location)] = population_freq\n",
    "                            \n",
    "    file.close()\n",
    "    \n",
    "    return population_freqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_name = \"Akkermansia_muciniphila_55290\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "parse_population_freqs(species_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "parse_population_freqs(species_name, use_HMP_freqs=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Conclusion:** `parse_population_freqs(species_name, use_HMP_freqs=True)` in `calculate_snp_prevalences` does work!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 2: does `calculate_within_persons_sfs.py` successfully create a within person sfs using HMP snp prevalences?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import parse_midas_data\n",
    "import pylab\n",
    "import sys\n",
    "import numpy\n",
    "import bz2\n",
    "import calculate_snp_prevalences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_name = \"Akkermansia_muciniphila_55290\"\n",
    "debug = False\n",
    "chunk_size = 1000000000"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Should we do this? \n",
    "sys.stderr.write(\"Loading core genes...\\n\")\n",
    "core_genes = parse_midas_data.load_core_genes(species_name)\n",
    "sys.stderr.write(\"Done! %d core genes\\n\" % len(core_genes))\n",
    "allowed_genes = core_genes\n",
    "\n",
    "sys.stderr.write(\"Loading population freqs...\\n\")\n",
    "population_freqs = calculate_snp_prevalences.parse_population_freqs(species_name, use_HMP_freqs=True)\n",
    "sys.stderr.write(\"Done! %d SNVs\\n\" % len(population_freqs))\n",
    "\n",
    "allowed_variant_type_list = ['1D','2D','3D','4D']\n",
    "allowed_variant_types = set(allowed_variant_type_list)  \n",
    "\n",
    "# Open post-processed MIDAS output\n",
    "snp_file =  bz2.BZ2File(\"%ssnps/%s/annotated_snps.txt.bz2\" % (parse_midas_data.data_directory, species_name),\"r\")\n",
    "    \n",
    "line = snp_file.readline() # header\n",
    "items = line.split()[1:]\n",
    "samples = numpy.array([item.strip() for item in items])\n",
    "\n",
    "# We shouldn't be doing this for raw data \n",
    "#samples = parse_midas_data.parse_merged_sample_names(items)\n",
    "    \n",
    "site_map = [{} for sample in samples]\n",
    "for sample_idx in xrange(0,len(samples)):\n",
    "    site_map[sample_idx] = {variant_type:{} for variant_type in allowed_variant_types}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "sys.stderr.write(\"Calculating within-person SFSs...\\n\")        \n",
    "num_sites_processed = 0\n",
    "for line in snp_file:\n",
    "    #\n",
    "    items = line.split()\n",
    "    # Load information about site\n",
    "    info_items = items[0].split(\"|\")\n",
    "    chromosome = info_items[0]\n",
    "    location = long(info_items[1])\n",
    "    gene_name = info_items[2]\n",
    "    variant_type = info_items[3]\n",
    "    \n",
    "    if len(info_items) > 5: # for backwards compatability\n",
    "            polarization = info_items[4]\n",
    "            pvalue = float(info_items[5])\n",
    "    else: \n",
    "        polarization=\"?\"\n",
    "        pvalue = float(info_items[4])\n",
    "        \n",
    "    #    \n",
    "    if variant_type not in allowed_variant_types:\n",
    "        continue\n",
    "    #    \n",
    "    if len(allowed_genes)>0 and (gene_name not in allowed_genes):\n",
    "        continue\n",
    "    #    \n",
    "    # Load alt and depth counts\n",
    "    alts = []\n",
    "    depths = []\n",
    "    for item in items[1:]:\n",
    "        subitems = item.split(\",\")\n",
    "        alts.append(long(subitems[0]))\n",
    "        depths.append(long(subitems[1]))\n",
    "    alts = numpy.array(alts)\n",
    "    depths = numpy.array(depths)\n",
    "    refs = depths-alts\n",
    "    #print alts\n",
    "    #print depths\n",
    "    #\n",
    "    # population_freq returns the fraction of people for which the alt is the major allele.\n",
    "    # This is a very important quantity being computed! It is later used for identifying CPS samples.\n",
    "    if (chromosome, location) in population_freqs:\n",
    "        population_freq = population_freqs[(chromosome, location)]\n",
    "    else:\n",
    "        population_freq = 0\n",
    "    \n",
    "    # polarize SFS according to population freq\n",
    "    if population_freq>0.5:\n",
    "        alts,refs = refs,alts\n",
    "        population_freq = 1-population_freq\n",
    "        \n",
    "    #    \n",
    "    for i in xrange(0,len(alts)):\n",
    "        site = (depths[i],alts[i])\n",
    "        #\n",
    "        if site not in site_map[i][variant_type]:\n",
    "            site_map[i][variant_type][site] = [0,0.0]\n",
    "        #        \n",
    "        site_map[i][variant_type][site][0] += 1\n",
    "        site_map[i][variant_type][site][1] += population_freq # weight of polarization reversals\n",
    "        #\n",
    "        #\n",
    "    num_sites_processed+=1\n",
    "    #print num_sites_processed\n",
    "    if num_sites_processed%50000==0:\n",
    "        sys.stderr.write(\"%dk sites processed...\\n\" % (num_sites_processed/1000))   \n",
    "        if debug:\n",
    "            break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "site_map[0]['1D'][(169, 0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sorted(site_map[0]['1D'].keys())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Conclusion:** It successfully calculates different within person SFSs based on setting `use_HMP_freqs=True`\n",
    "- Furthermore, I confirmed that the within_host_sfs.txt.bz2 file is indeed the product of the pipeline when using `use_HMP_freqs=True`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Question 3: What are the within and between difference values for each sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.insert(0, \"/u/home/m/michaelw/project-ngarud/microbiome_evolution/microbiome_evolution_MOUSE/\")\n",
    "\n",
    "import numpy\n",
    "from scipy.linalg import eigh\n",
    "from scipy.cluster.hierarchy import dendrogram, linkage\n",
    "from scipy.cluster.hierarchy import cophenet\n",
    "from scipy.cluster.hierarchy import fcluster\n",
    "from numpy.random import shuffle, normal\n",
    "import scipy.stats\n",
    "from scipy.stats import binom\n",
    "import config\n",
    "from scipy.special import betainc\n",
    "import parse_midas_data\n",
    "import sample_utils\n",
    "import stats_utils\n",
    "import os.path\n",
    "import sfs_utils\n",
    "import diversity_utils\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_name = \"Akkermansia_muciniphila_55290\"\n",
    "desired_samples = diversity_utils.calculate_highcoverage_samples(species_name, 20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples, sfs_map = parse_midas_data.parse_within_sample_sfs(species_name,     allowed_variant_types=set(['4D']))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "total_vec = []\n",
    "within_vec = []\n",
    "between_vec= []\n",
    "betweenX10percent = []\n",
    "for sample in samples:\n",
    "    within_sites, between_sites, total_sites = sfs_utils.calculate_polymorphism_rates_from_sfs_map(sfs_map[sample])\n",
    "    \n",
    "    total_vec.append(int(total_sites))\n",
    "    within_vec.append(int(within_sites))\n",
    "    between_vec.append(int(between_sites))\n",
    "    betweenX10percent.append(0.1*between_sites)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.DataFrame()\n",
    "df['samples'] = samples\n",
    "df['total_sites'] = total_vec\n",
    "df['within_sites'] = within_vec\n",
    "df['between_sites'] = between_vec\n",
    "df['betweenX10percent'] = betweenX10percent"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>samples</th>\n",
       "      <th>total_sites</th>\n",
       "      <th>within_sites</th>\n",
       "      <th>between_sites</th>\n",
       "      <th>betweenX10percent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M8CeCGG</td>\n",
       "      <td>274399</td>\n",
       "      <td>19</td>\n",
       "      <td>8169</td>\n",
       "      <td>816.966868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M8JGG_2</td>\n",
       "      <td>274582</td>\n",
       "      <td>17</td>\n",
       "      <td>8219</td>\n",
       "      <td>821.990907</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>M8IGG_2</td>\n",
       "      <td>274676</td>\n",
       "      <td>18</td>\n",
       "      <td>8218</td>\n",
       "      <td>821.852050</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>M7CecGG</td>\n",
       "      <td>272344</td>\n",
       "      <td>20</td>\n",
       "      <td>8132</td>\n",
       "      <td>813.262300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>M8ColonGG_2</td>\n",
       "      <td>274399</td>\n",
       "      <td>21</td>\n",
       "      <td>8206</td>\n",
       "      <td>820.636318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>M7IGG_2</td>\n",
       "      <td>274628</td>\n",
       "      <td>19</td>\n",
       "      <td>8216</td>\n",
       "      <td>821.679176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>TL1gDNAshort</td>\n",
       "      <td>274714</td>\n",
       "      <td>17</td>\n",
       "      <td>8215</td>\n",
       "      <td>821.576558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>M7JGG</td>\n",
       "      <td>274468</td>\n",
       "      <td>18</td>\n",
       "      <td>8213</td>\n",
       "      <td>821.364061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>M7DGG</td>\n",
       "      <td>273872</td>\n",
       "      <td>23</td>\n",
       "      <td>8137</td>\n",
       "      <td>813.734558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>M8D</td>\n",
       "      <td>274627</td>\n",
       "      <td>18</td>\n",
       "      <td>8217</td>\n",
       "      <td>821.791789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>M7ColonGG</td>\n",
       "      <td>274436</td>\n",
       "      <td>20</td>\n",
       "      <td>8195</td>\n",
       "      <td>819.538913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>M1JC_CKDN220050970-1A_H7MMHDSX5_L2</td>\n",
       "      <td>273760</td>\n",
       "      <td>20</td>\n",
       "      <td>8129</td>\n",
       "      <td>812.910320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>M1CeC_CKDN220050952-1A_H7MMHDSX5_L3</td>\n",
       "      <td>273581</td>\n",
       "      <td>17</td>\n",
       "      <td>8151</td>\n",
       "      <td>815.195402</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>M1CoC_CKDN220050958-1A_H7MMHDSX5_L2</td>\n",
       "      <td>273131</td>\n",
       "      <td>21</td>\n",
       "      <td>8079</td>\n",
       "      <td>807.982685</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>M1IC_CKDN220050976-1A_H7MMHDSX5_L2</td>\n",
       "      <td>273844</td>\n",
       "      <td>24</td>\n",
       "      <td>8130</td>\n",
       "      <td>813.002262</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>M1DC_CKDN220050964-1A_H7MMHDSX5_L2</td>\n",
       "      <td>273762</td>\n",
       "      <td>20</td>\n",
       "      <td>8117</td>\n",
       "      <td>811.771460</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>M2IC_CKDN220050977-1A_H7MMHDSX5_L2</td>\n",
       "      <td>274225</td>\n",
       "      <td>20</td>\n",
       "      <td>8176</td>\n",
       "      <td>817.600858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>M3CeC_CKDN220050954-1A_H7MMHDSX5_L2</td>\n",
       "      <td>272070</td>\n",
       "      <td>27</td>\n",
       "      <td>8032</td>\n",
       "      <td>803.269843</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>M6IG_CKDN220050981-1A_HF3V5DSX5_L1</td>\n",
       "      <td>273914</td>\n",
       "      <td>20</td>\n",
       "      <td>8143</td>\n",
       "      <td>814.301862</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>M2CeC_CKDN220050953-1A_merge</td>\n",
       "      <td>272975</td>\n",
       "      <td>19</td>\n",
       "      <td>8058</td>\n",
       "      <td>805.848495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>M2CoC_CKDN220050959-1A_H7MMHDSX5_L2</td>\n",
       "      <td>272865</td>\n",
       "      <td>22</td>\n",
       "      <td>8066</td>\n",
       "      <td>806.615714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>M3CoC_CKDN220050960-1A_H7MMHDSX5_L2</td>\n",
       "      <td>273472</td>\n",
       "      <td>23</td>\n",
       "      <td>8086</td>\n",
       "      <td>808.670205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>M6JG_CKDN220050975-1A_H7MMHDSX5_L2</td>\n",
       "      <td>274102</td>\n",
       "      <td>18</td>\n",
       "      <td>8137</td>\n",
       "      <td>813.739381</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>M6CeG_CKDN220050957-1A_H7MMGDSX5_L2</td>\n",
       "      <td>273981</td>\n",
       "      <td>23</td>\n",
       "      <td>8159</td>\n",
       "      <td>815.988449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>M4CoG_CKDN220050961-1A_merge</td>\n",
       "      <td>273534</td>\n",
       "      <td>19</td>\n",
       "      <td>8101</td>\n",
       "      <td>810.153076</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>M3JC_CKDN220050972-1A_H7MMHDSX5_L2</td>\n",
       "      <td>273983</td>\n",
       "      <td>21</td>\n",
       "      <td>8123</td>\n",
       "      <td>812.355751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>M6CoG_CKDN220050963-1A_H7MMHDSX5_L2</td>\n",
       "      <td>273690</td>\n",
       "      <td>21</td>\n",
       "      <td>8133</td>\n",
       "      <td>813.319202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>M2JC_CKDN220050971-1A_H7MMHDSX5_L2</td>\n",
       "      <td>273968</td>\n",
       "      <td>19</td>\n",
       "      <td>8137</td>\n",
       "      <td>813.744140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>M5CeG_CKDN220050956-1A_H7MMHDSX5_L2</td>\n",
       "      <td>273188</td>\n",
       "      <td>22</td>\n",
       "      <td>8082</td>\n",
       "      <td>808.227644</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>M5CoG_CKDN220050962-1A_merge</td>\n",
       "      <td>273623</td>\n",
       "      <td>21</td>\n",
       "      <td>8107</td>\n",
       "      <td>810.717275</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>M4CeG_CKDN220050955-1A_H7MMHDSX5_L2</td>\n",
       "      <td>273795</td>\n",
       "      <td>20</td>\n",
       "      <td>8142</td>\n",
       "      <td>814.239928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>M5IG_CKDN220050980-1A_HF3V5DSX5_L1</td>\n",
       "      <td>273985</td>\n",
       "      <td>18</td>\n",
       "      <td>8146</td>\n",
       "      <td>814.603800</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>M4IG_CKDN220050979-1A_H7MMHDSX5_L2</td>\n",
       "      <td>273869</td>\n",
       "      <td>22</td>\n",
       "      <td>8129</td>\n",
       "      <td>812.902448</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>M4DG_CKDN220050967-1A_H7MMHDSX5_L3</td>\n",
       "      <td>273796</td>\n",
       "      <td>16</td>\n",
       "      <td>8151</td>\n",
       "      <td>815.192558</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>M4JG_CKDN220050973-1A_merge</td>\n",
       "      <td>274319</td>\n",
       "      <td>20</td>\n",
       "      <td>8188</td>\n",
       "      <td>818.800692</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>M5JG_CKDN220050974-1A_merge</td>\n",
       "      <td>273937</td>\n",
       "      <td>18</td>\n",
       "      <td>8144</td>\n",
       "      <td>814.409653</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>M5DG_CKDN220050968-1A_merge</td>\n",
       "      <td>273883</td>\n",
       "      <td>20</td>\n",
       "      <td>8147</td>\n",
       "      <td>814.738755</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>M3DC_CKDN220050966-1A_H7MMHDSX5_L2</td>\n",
       "      <td>273948</td>\n",
       "      <td>22</td>\n",
       "      <td>8183</td>\n",
       "      <td>818.364811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>M6DG_CKDN220050969-1A_H7MMGDSX5_L1</td>\n",
       "      <td>274119</td>\n",
       "      <td>24</td>\n",
       "      <td>8169</td>\n",
       "      <td>816.967622</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                samples  total_sites  within_sites  \\\n",
       "0                               M8CeCGG       274399            19   \n",
       "1                               M8JGG_2       274582            17   \n",
       "2                               M8IGG_2       274676            18   \n",
       "3                               M7CecGG       272344            20   \n",
       "4                           M8ColonGG_2       274399            21   \n",
       "5                               M7IGG_2       274628            19   \n",
       "6                          TL1gDNAshort       274714            17   \n",
       "7                                 M7JGG       274468            18   \n",
       "8                                 M7DGG       273872            23   \n",
       "9                                   M8D       274627            18   \n",
       "10                            M7ColonGG       274436            20   \n",
       "11   M1JC_CKDN220050970-1A_H7MMHDSX5_L2       273760            20   \n",
       "12  M1CeC_CKDN220050952-1A_H7MMHDSX5_L3       273581            17   \n",
       "13  M1CoC_CKDN220050958-1A_H7MMHDSX5_L2       273131            21   \n",
       "14   M1IC_CKDN220050976-1A_H7MMHDSX5_L2       273844            24   \n",
       "15   M1DC_CKDN220050964-1A_H7MMHDSX5_L2       273762            20   \n",
       "16   M2IC_CKDN220050977-1A_H7MMHDSX5_L2       274225            20   \n",
       "17  M3CeC_CKDN220050954-1A_H7MMHDSX5_L2       272070            27   \n",
       "18   M6IG_CKDN220050981-1A_HF3V5DSX5_L1       273914            20   \n",
       "19         M2CeC_CKDN220050953-1A_merge       272975            19   \n",
       "20  M2CoC_CKDN220050959-1A_H7MMHDSX5_L2       272865            22   \n",
       "21  M3CoC_CKDN220050960-1A_H7MMHDSX5_L2       273472            23   \n",
       "22   M6JG_CKDN220050975-1A_H7MMHDSX5_L2       274102            18   \n",
       "23  M6CeG_CKDN220050957-1A_H7MMGDSX5_L2       273981            23   \n",
       "24         M4CoG_CKDN220050961-1A_merge       273534            19   \n",
       "25   M3JC_CKDN220050972-1A_H7MMHDSX5_L2       273983            21   \n",
       "26  M6CoG_CKDN220050963-1A_H7MMHDSX5_L2       273690            21   \n",
       "27   M2JC_CKDN220050971-1A_H7MMHDSX5_L2       273968            19   \n",
       "28  M5CeG_CKDN220050956-1A_H7MMHDSX5_L2       273188            22   \n",
       "29         M5CoG_CKDN220050962-1A_merge       273623            21   \n",
       "30  M4CeG_CKDN220050955-1A_H7MMHDSX5_L2       273795            20   \n",
       "31   M5IG_CKDN220050980-1A_HF3V5DSX5_L1       273985            18   \n",
       "32   M4IG_CKDN220050979-1A_H7MMHDSX5_L2       273869            22   \n",
       "33   M4DG_CKDN220050967-1A_H7MMHDSX5_L3       273796            16   \n",
       "34          M4JG_CKDN220050973-1A_merge       274319            20   \n",
       "35          M5JG_CKDN220050974-1A_merge       273937            18   \n",
       "36          M5DG_CKDN220050968-1A_merge       273883            20   \n",
       "37   M3DC_CKDN220050966-1A_H7MMHDSX5_L2       273948            22   \n",
       "38   M6DG_CKDN220050969-1A_H7MMGDSX5_L1       274119            24   \n",
       "\n",
       "    between_sites  betweenX10percent  \n",
       "0            8169         816.966868  \n",
       "1            8219         821.990907  \n",
       "2            8218         821.852050  \n",
       "3            8132         813.262300  \n",
       "4            8206         820.636318  \n",
       "5            8216         821.679176  \n",
       "6            8215         821.576558  \n",
       "7            8213         821.364061  \n",
       "8            8137         813.734558  \n",
       "9            8217         821.791789  \n",
       "10           8195         819.538913  \n",
       "11           8129         812.910320  \n",
       "12           8151         815.195402  \n",
       "13           8079         807.982685  \n",
       "14           8130         813.002262  \n",
       "15           8117         811.771460  \n",
       "16           8176         817.600858  \n",
       "17           8032         803.269843  \n",
       "18           8143         814.301862  \n",
       "19           8058         805.848495  \n",
       "20           8066         806.615714  \n",
       "21           8086         808.670205  \n",
       "22           8137         813.739381  \n",
       "23           8159         815.988449  \n",
       "24           8101         810.153076  \n",
       "25           8123         812.355751  \n",
       "26           8133         813.319202  \n",
       "27           8137         813.744140  \n",
       "28           8082         808.227644  \n",
       "29           8107         810.717275  \n",
       "30           8142         814.239928  \n",
       "31           8146         814.603800  \n",
       "32           8129         812.902448  \n",
       "33           8151         815.192558  \n",
       "34           8188         818.800692  \n",
       "35           8144         814.409653  \n",
       "36           8147         814.738755  \n",
       "37           8183         818.364811  \n",
       "38           8169         816.967622  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "python27_env",
   "language": "python",
   "name": "python27_env"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
