{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import lz4.frame\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "sys.path.append(\"/u/project/ngarud/michaelw/Diversity-Along-Gut/ConventionalMouse/scripts/postprocessing/postprocessing_scripts/\")\n",
    "import config\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Creating strainfinder input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_samples = np.loadtxt(config.accessions, dtype=str, comments=\"#\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_name = \"100042\"\n",
    "data_directory = config.data_directory\n",
    "snps_summary_file = pd.read_csv(\"%ssnps/snps_summary.tsv\" % (data_directory),sep=\"\\t\",index_col=0) # MW 09/11/2025: new way for parsing sample_order\n",
    "snps_summary_file = snps_summary_file[snps_summary_file[\"species_id\"] == int(species_name)] # MW 09/11/2025: new way for parsing sample_order\n",
    "sample_order = snps_summary_file.index # MW 09/11/2025: new way for parsing sample_order\n",
    "sample_order = np.array(list(sample_order)) # MW 09/11/2025: new way for parsing sample_order"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## `diversity_utils.py`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from diversity_utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Cec_1_1', 'Cec_1_2', 'Cec_1_3', 'Cec_2_1', 'Cec_2_2', 'Cec_2_3',\n",
       "       'Cec_3_1', 'Cec_3_2', 'Cec_3_3', 'Col_1_1', 'Col_1_2', 'Col_1_3',\n",
       "       'Col_2_1', 'Col_2_2', 'Col_2_3', 'Col_3_1', 'Col_3_2', 'Col_3_3'],\n",
       "      dtype='<U7')"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_name = \"251524\"\n",
    "highcoverage_samples = calculate_highcoverage_samples(species_name)\n",
    "highcoverage_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Cec_1_1', 'Cec_1_2', 'Cec_1_3', 'Cec_2_1', 'Cec_2_2', 'Cec_2_3',\n",
       "       'Cec_3_1', 'Cec_3_2', 'Cec_3_3', 'Col_1_1', 'Col_1_2', 'Col_1_3',\n",
       "       'Col_2_1', 'Col_2_2', 'Col_2_3', 'Col_3_1', 'Col_3_2', 'Col_3_3'],\n",
       "      dtype='<U7')"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "species_name = \"251524\"\n",
    "haploid_samples = calculate_haploid_samples(species_name, quick_and_dirty = True)\n",
    "haploid_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### division by 0 error in `calculate_haploid_samples()`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sfs_utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_name = \"251524\"\n",
    "min_coverage=config.min_median_coverage\n",
    "threshold_pi=config.threshold_pi\n",
    "threshold_within_between_fraction=config.threshold_within_between_fraction\n",
    "debug=False\n",
    "quick_and_dirty=True\n",
    "quick_and_dirty_threshold = 5e-4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "desired_samples = calculate_highcoverage_samples(species_name, min_coverage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples, sfs_map = parse_midas_data.parse_within_sample_sfs(species_name, allowed_variant_types=set(['4D']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sfs_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Why can I not make within-sample sfs for some species?\n",
    "Example species: 251524"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import parse_midas_data\n",
    "import pylab\n",
    "import sys\n",
    "import numpy\n",
    "import bz2\n",
    "import calculate_snp_prevalences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_name = \"251524\"\n",
    "debug = True\n",
    "chunk_size = 1000000000\n",
    "use_HMP = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "core_genes = parse_midas_data.load_core_genes(species_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "core_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Core gene utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy\n",
    "import sys\n",
    "sys.path.append(\"/u/project/ngarud/michaelw/Diversity-Along-Gut/ConventionalMouse/scripts/postprocessing/postprocessing_scripts/\")\n",
    "import config\n",
    "import gzip\n",
    "import os.path\n",
    "import os\n",
    "import midas_db_utils\n",
    "import parse_midas_data #MW: Added 09/21/23\n",
    "\n",
    "from core_gene_utils import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "core_genes_directory = (\"%score_genes/\" % config.data_directory)\n",
    "external_core_genes_directory = (\"%score_genes/external/\" % config.data_directory)\n",
    "\n",
    "default_external_shared_gene_filename = (external_core_genes_directory+\"shared_genes.txt.gz\")\n",
    "default_external_core_gene_filename = (external_core_genes_directory+\"core_genes.txt.gz\")\n",
    "default_external_stringent_core_gene_filename = (external_core_genes_directory+\"core_genes_stringent.txt.gz\")\n",
    "default_external_gene_freq_template = (external_core_genes_directory+\"%s_gene_freqs.txt.gz\")\n",
    "\n",
    "default_shared_gene_filename = (core_genes_directory+\"shared_genes.txt.gz\")\n",
    "default_core_gene_filename = (core_genes_directory+\"core_genes.txt.gz\")\n",
    "default_stringent_core_gene_filename = (core_genes_directory+\"core_genes_stringent.txt.gz\")\n",
    "default_gene_freq_template = (core_genes_directory+\"%s_gene_freqs.txt.gz\")\n",
    "\n",
    "#HMP data #MW added on 11/2/23\n",
    "shared_gene_HMP_filename = (core_genes_directory+\"shared_genes_HMP.txt.gz\")\n",
    "core_gene_HMP_filename = (core_genes_directory+\"core_genes_HMP.txt.gz\")\n",
    "stringent_core_gene_HMP_filename = (core_genes_directory+\"core_genes_stringent_HMP.txt.gz\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "use_HMP = False\n",
    "\n",
    "import parse_midas_data\n",
    "\n",
    "# os.system('mkdir -p %s' % core_genes_directory)\n",
    "# os.system('mkdir -p %s' % external_core_genes_directory)\n",
    "\n",
    "# pangenome_species = parse_midas_data.parse_good_species_list()\n",
    "pangenome_species = parse_midas_data.parse_species_list() # MW 08/29/2025: loaded all species in the species_snps.txt file\n",
    "\n",
    "cmin = config.core_genome_min_copynum\n",
    "cmax = config.core_genome_max_copynum  \n",
    "shared_cmin = config.shared_genome_min_copynum\n",
    "\n",
    "min_good_fraction = config.core_genome_min_prevalence\n",
    "min_coverage = 5 # (for assessing core genome, we'll use a lower coverage value than when we look at real changes)\n",
    "\n",
    "#Loading files to write to\n",
    "output_filename = default_core_gene_filename \n",
    "# output_file = gzip.open(output_filename, mode = \"wt\", encoding = \"utf-8\") # MW 07/09/2025: changing gzip.GzipFile to gzip open, mode to \"wt\", and added encoding = \"utf-8\"\n",
    "\n",
    "stringent_output_filename = default_stringent_core_gene_filename\n",
    "# stringent_output_file = gzip.open(stringent_output_filename,mode = \"wt\", encoding=\"utf-8\") # MW 07/09/2025: changing gzip.GzipFile to gzip open, mode to \"wt\", and added encoding = \"utf-8\"\n",
    "\n",
    "# shared_output_file = gzip.open(default_shared_gene_filename, mode = \"wt\", encoding=\"utf-8\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_name = \"251524\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load reference genes\n",
    "sys.stderr.write(\"Loading genes on reference genome..\\n\")\n",
    "reference_genes = midas_db_utils.load_reference_genes(species_name)\n",
    "sys.stderr.write(\"Done!\\n\")\n",
    "\n",
    "# Load reference genes\n",
    "sys.stderr.write(\"Loading shared genes from midas db..\\n\")\n",
    "# midas_shared_genes = midas_db_utils.parse_midas_shared_genes(species_name)  # MW 07/08/2025: no cross_species_centroids.txt.gz file            \n",
    "sys.stderr.write(\"Done!\\n\")\n",
    "\n",
    "\n",
    "bad_pangenome_data = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load gene coverage information for species_name\n",
    "sys.stderr.write(\"Loading pangenome data for %s...\\n\" % species_name)\n",
    "gene_samples, gene_names, gene_presence_matrix, gene_depth_matrix, marker_coverages, gene_reads_matrix = parse_midas_data.parse_pangenome_data(species_name)\n",
    "sys.stderr.write(\"Done!\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "high_coverage_idxs = (marker_coverages>=min_coverage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "high_coverage_idxs.sum() < 0.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_names = numpy.array(gene_names)\n",
    "gene_samples = gene_samples[high_coverage_idxs]\n",
    "marker_coverages = marker_coverages[high_coverage_idxs]\n",
    "gene_depth_matrix = gene_depth_matrix[:,high_coverage_idxs] \n",
    "gene_copynum_matrix = gene_depth_matrix*1.0/(marker_coverages+(marker_coverages==0))\n",
    "\n",
    "good_sample_idxs = get_good_pangenome_samples(species_name, marker_coverages, gene_copynum_matrix) # MW 09/03/2025: added species name\n",
    "bad_sample_idxs = numpy.logical_not(good_sample_idxs)\n",
    "        \n",
    "#sys.stderr.write(\"%d bad samples!\\n\" % bad_sample_idxs.sum())\n",
    "\n",
    "gene_samples = gene_samples[good_sample_idxs]\n",
    "marker_coverages = marker_coverages[good_sample_idxs]\n",
    "gene_copynum_matrix = gene_copynum_matrix[:,good_sample_idxs]\n",
    "\n",
    "reference_gene_idxs = numpy.array([gene_name in reference_genes for gene_name in gene_names])\n",
    "\n",
    "# midas_shared_idxs = numpy.array([gene_name in midas_shared_genes for gene_name in gene_names]) # MW 07/09/2025: no cross_species_centroids.txt.gz\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "reference_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "species_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# These are genes that have coverage >=3x normal in some sample. This are candidates for being linked to another species.\n",
    "# (they could also be multi-copy genes, but we can't look at much on these genes anyway, so might as well toss them out)\n",
    "metagenome_shared_idxs = ((gene_copynum_matrix>shared_cmin).sum(axis=1)>0.5)\n",
    "\n",
    "# Now union with those we identified from midas db\n",
    "# shared_idxs = numpy.logical_or(metagenome_shared_idxs, midas_shared_idxs) \n",
    "shared_idxs = metagenome_shared_idxs # MW 07/09/2025: no cross_species_centroids.txt.gz\n",
    "non_shared_idxs = numpy.logical_not(shared_idxs)\n",
    "\n",
    "shared_gene_names = gene_names[shared_idxs]\n",
    "\n",
    "# calculating good genes\n",
    "good_idxs = (((gene_copynum_matrix>=cmin)*(gene_copynum_matrix<=cmax)).sum(axis=1)*1.0/len(marker_coverages) >= min_good_fraction) #indexes in which gene copy numbers are between 0.3 and 3 in 90% of samples\n",
    "core_gene_idxs = good_idxs*reference_gene_idxs*non_shared_idxs\n",
    "core_gene_names = gene_names[core_gene_idxs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sum(reference_gene_idxs)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "py38",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.20"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
